rcl_action/include/rcl_action/action_client.h

```diff
@@ -0,0 +1,690 @@
+ * be allocated for a dynamically sized array in the target message, then the
+ * allocator given in the action client options is used.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Maybe [1]
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ * <i>[1] only if required when filling the result response message, avoided for fixed sizes</i>
+ *
+ * \param[in] action_client handle to the client that will take the result response
+ * \param[out] ros_result_response preallocated struct where the ROS result message is copied.
+ * \return `RCL_RET_OK` if the response was taken successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_CLIENT_INVALID` if the action client is invalid, or
+ * \return `RCL_RET_BAD_ALLOC` if allocating memory failed, or
+ * \return `RCL_RET_ACTION_CLIENT_TAKE_FAILED` if take failed but no error occurred
+ *         in the middleware, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_take_result_response(
+  const rcl_action_client_t * action_client,
+  void * ros_result);
+
+/// Send a cancel request for a goal using a rcl_action_client_t.
+/**
+ * This is a non-blocking call.
+ *
+ * The caller is responsible for ensuring that the type of `ros_cancel_request`
+ * and the type associate with the client (via the type support) match.
+ * Passing a different type produces undefined behavior and cannot be checked
+ * by this function and therefore no deliberate error will occur.
+ *
+ * The following cancel policy applies based on the goal ID and the timestamp provided
+ * by the `ros_cancel_request` message:
+ *
+ * - If the goal ID is zero and timestamp is zero, cancel all goals.
+ * - If the goal ID is zero and timestamp is not zero, cancel all goals accepted
+ *   at or before the timestamp.
+ * - If the goal ID is not zero and timestamp is zero, cancel the goal with the
+ *   given ID regardless of the time it was accepted.
+ * - If the goal ID is not zero and timestamp is not zero, cancel the goal with the
+ *   given ID and all goals accepted at or before the timestamp.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_client handle to the client that will make the cancel request
+ * \param[in] ros_cancel_request pointer the ROS cancel request message
+ * \return `RCL_RET_OK` if the response was taken successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_CLIENT_INVALID` if the action client is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_cancel_request(
```

Comment:
@jacobperron same about exposing `rmw_request_id_t`

Reply:
Assuming we have a goal ID as part of the ROS request message, I'm not sure that we need to expose a `rmw_request_id_t`.